packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp

```diff
@@ -475,39 +476,48 @@ void RebalanceCoarseCompositeOperator(const int maxRegPerProc,
 {
 #include "MueLu_UseShortNames.hpp"
   using CoordType = typename Teuchos::ScalarTraits<Scalar>::coordinateType;
+  using Teuchos::TimeMonitor;
+
+  RCP<TimeMonitor> tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("ReblanceCoarseCompositeOperator: ")));
 
   RCP<Teuchos::FancyOStream> fos = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
   fos->setOutputToRootOnly(0);
   *fos << "Rebalancing coarse composite operator." << std::endl;
 
-  //RCP<const Map> map = coarseCompOp->getMap();
-
   const int numPartitions = rebalanceNumPartitions;
 
   // We build a fake level
   Level level;
-  level.SetLevelID(1);
+  level.SetLevelID(3);
 
   RCP<FactoryManagerBase> factoryHandler = rcp(new FactoryManager());
   level.SetFactoryManager(factoryHandler);
 
+  RCP<TimeMonitor> tmLocal = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("rebalanceCoarse: Zoltan construction:")));
+
   RCP<Zoltan2Interface> zoltan = rcp(new Zoltan2Interface());
 
   level.Set<RCP<Matrix> >     ("A",                    coarseCompOp);
   level.Set<RCP<MultiVector> >("Coordinates",          compCoarseCoordinates);
-  level.Set<int>              ("number of partitions", numPartitions);
 
   RCP<RepartitionFactory> repart = rcp(new RepartitionFactory());
   Teuchos::ParameterList paramList;
   paramList.set("repartition: remap parts",       false);
+  if( numPartitions > 0 ){ // If number of coarse rebalance partitions was provided by the user.
+    level.Set<int>              ("number of partitions", numPartitions);
+  } else {
+    RCP<RepartitionHeuristicFactory> repartHeuristic = rcp(new RepartitionHeuristicFactory());
```

Comment:
Don't you need to pass in parameters somewhere?

Reply:
How do I do that? Or which parameter list would that be?

Reply:
Did you check that you can actually set the parameters that the heuristic factory takes? Cause otherwise, you'll just be running the defaults all the time.

Reply:
You could have a look at how the `ParameterListInterpreter` handles the parameters.